            Medical Appointment No Shows Dataset(https://www.kaggle.com/joniarroba/noshowappointments)
1, The 'Medical Appointment No Shows' dataset includes 110k medical appointments and its 15 variables (characteristics) of each, like age, gender,and if the patient has financial supports. The target is the last column which shows if the patient show-up or no-show the appointment.The task is to predict if someone would not show up for an appointment.

2, load the dataset

In [82]:
import pandas as pd
import numpy as np
import math
from sklearn import tree as tree
from sklearn.metrics import classification_report
from sklearn.feature_selection import SelectFromModel
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, mean_squared_error

In [83]:
df = pd.read_csv('KaggleV2-May-2016.csv')
print(df.shape)
df.head()

(110527, 14)


PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   

         AppointmentDay  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No  
1         0           0        0             0      No  
2         0           0        0             0      No  
3         0           0        0             0      No  
4         1           0        0             0      No

The dataset has 110527 instances and 14 features, some features are helpless for us to make predictions, like AppointmentID, Appointment time, those features will be dropped during the learning process.

For the 'Gender' attribute, change the 'F' into 1, 'M' into 0; for the label, change 'Yes' into 1, 'No' into 0.

In [84]:
df = df.replace(['F','M','Yes','No'],[1,0,1,0])

In [85]:
y = df['No-show']
df_new = df.drop(['PatientId','AppointmentID','ScheduledDay','AppointmentDay','Neighbourhood','No-show'],axis=1)
df_new.head()

Gender  Age  Scholarship  Hipertension  Diabetes  Alcoholism  Handcap  \
0       1   62            0             1         0           0        0   
1       0   56            0             0         0           0        0   
2       1   62            0             0         0           0        0   
3       1    8            0             0         0           0        0   
4       1   56            0             1         1           0        0   

   SMS_received  
0             0  
1             0  
2             0  
3             0  
4             0

Because the training and test sets have not been given, we need to split those two sets first.

In [86]:
X_train,X_test,y_train,y_test=train_test_split(df_new,y,test_size=0.333,random_state=123)
print(X_train.shape,y_train.shape,X_test.shape,y_test.shape)

(73721, 8) (73721,) (36806, 8) (36806,)


3, Train a decision tree of at most depth = 1, visualize it, test it on train, and test it on test. For the criterion, I use entropy.

In [87]:
dt_model_1 = tree.DecisionTreeClassifier(criterion='entropy',max_depth=1,random_state=123)

In [88]:
dt_model_1 = dt_model_1.fit(X_train,y_train)

In [89]:
feature_list=X_train.columns
with open('tree1.dot','w+') as f:
    f=tree.export_graphviz(dt_model_1, feature_names = feature_list, out_file=f)

The printed tree with 1 depth is in the 'DT with 1 depth.png'.

Test the tree on the train.

In [90]:
y_1_train= dt_model_1.predict(X_train)
print('Accuracy score on the training set: ', accuracy_score(y_train, y_1_train))

Accuracy score on the training set:  0.799175268919


Test the tree on the test.

In [91]:
y_1_test = dt_model_1.predict(X_test)
print('Accuracy score on the test set: ', accuracy_score(y_test, y_1_test))

Accuracy score on the test set:  0.795848502961


4, Train a decision tree of at most depth = 2, visualize it, test it on train, and test it on test. For the criterion, I use entropy.

In [92]:
dt_model_2 = tree.DecisionTreeClassifier(criterion='entropy',max_depth=2,random_state=123)
dt_model_2 = dt_model_2.fit(X_train,y_train)
with open('tree2.dot','w+') as f2:
    f2=tree.export_graphviz(dt_model_2, feature_names = feature_list, out_file=f2)

The printed tree with 2 depth is in the 'DT with 2 depth.png'.
Test the tree on the train. Then test the tree on the test.

In [93]:
y_2_train= dt_model_2.predict(X_train)
print('Accuracy score on the training set: ', accuracy_score(y_train, y_2_train))
y_2_test = dt_model_2.predict(X_test)
print('Accuracy score on the test set: ', accuracy_score(y_test, y_2_test))

Accuracy score on the training set:  0.799175268919
Accuracy score on the test set:  0.795848502961


5, Train a decision tree of at most depth = 3, visualize it, test it on train, and test it on test. For the criterion, I use entropy entropy.

In [94]:
dt_model_3 = tree.DecisionTreeClassifier(criterion='entropy',max_depth=3,random_state=123)
dt_model_3 = dt_model_3.fit(X_train,y_train)
with open('tree3.dot','w+') as f3:
    f3=tree.export_graphviz(dt_model_3, feature_names = feature_list, out_file=f3)

The printed tree with 3 depth is in the 'DT with 3 depth.png'. Test the tree on the train. Then test the tree on the test.

In [95]:
y_3_train= dt_model_3.predict(X_train)
print('Accuracy score on the training set: ', accuracy_score(y_train, y_3_train))
y_3_test = dt_model_3.predict(X_test)
print('Accuracy score on the test set: ', accuracy_score(y_test, y_3_test))

Accuracy score on the training set:  0.799175268919
Accuracy score on the test set:  0.795848502961


6, Train a decision tree of at most depth = 4, test it on train, and test it on test. For the criterion, I use entropy entropy.

In [96]:
dt_model_4 = tree.DecisionTreeClassifier(criterion='entropy',max_depth=4,random_state=123)
dt_model_4 = dt_model_4.fit(X_train,y_train)
y_4_train= dt_model_4.predict(X_train)
print('Accuracy score on the training set: ', accuracy_score(y_train, y_4_train))
y_4_test = dt_model_4.predict(X_test)
print('Accuracy score on the test set: ', accuracy_score(y_test, y_4_test))

Accuracy score on the training set:  0.799175268919
Accuracy score on the test set:  0.795848502961


7, Train a decision tree of at most depth = 5,test it on train, and test it on test. For the criterion, I use entropy entropy.

In [97]:
dt_model_5 = tree.DecisionTreeClassifier(criterion='entropy',max_depth=5,random_state=123)
dt_model_5 = dt_model_5.fit(X_train,y_train)
y_5_train= dt_model_5.predict(X_train)
print('Accuracy score on the training set: ', accuracy_score(y_train, y_5_train))
y_5_test = dt_model_5.predict(X_test)
print('Accuracy score on the test set: ', accuracy_score(y_test, y_5_test))

Accuracy score on the training set:  0.799188833575
Accuracy score on the test set:  0.795848502961


8, Train a decision tree of at most depth = 6,test it on train, and test it on test. For the criterion, I use entropy entropy.

In [98]:
dt_model_6 = tree.DecisionTreeClassifier(criterion='entropy',max_depth=5,random_state=123)
dt_model_6 = dt_model_6.fit(X_train,y_train)
y_6_train= dt_model_6.predict(X_train)
print('Accuracy score on the training set: ', accuracy_score(y_train, y_6_train))
y_6_test = dt_model_6.predict(X_test)
print('Accuracy score on the test set: ', accuracy_score(y_test, y_6_test))

Accuracy score on the training set:  0.799188833575
Accuracy score on the test set:  0.795848502961


Analysis: 
The split condition of the tree with 1 depth is if the patient received SMS before the scheduled date. The logic is quite reasonable, patients had not been reminded before the scheduled date and they are likely to forget the appointment. The accuracy based on test set is 0.795848502961, this prediction result is pretty good even if the tree only has one depth. The second split condition is the range of age, the chosen attribute makes sense, senior people are more likely to have a poor memory, they had forgotten the appointment. But the result is disappointing, the age attribute does not boost the result. The third split attribute is still age, the performance has not been improved.
Set the max_depth= 4,5,6, the prediction results has not been boosted at all, however, the result based on training set has been improved slightly, it increases from 0.799175268919 to 0.799188833575. The decison tree approach is not quite suitable for this dataset,the dataframe is too sparse. Except the 'Age' attribute, the most of the rest values of attributes are 0, the tree cannot pick up effective attributes to make splits.  
